pkg/domain/ru_stats.go

```diff
@@ -0,0 +1,279 @@
+func GetLastExpectedTime(now time.Time, interval time.Duration) time.Time {
+	nowTs := now.Unix()
+	intervalSecs := int64(interval / time.Second)
+	tzOffset := time.Date(1971, 1, 1, 0, 0, 0, 0, time.Local).Unix()
+	targetTs := nowTs - (nowTs+intervalSecs-tzOffset)%intervalSecs
+	return time.Unix(targetTs, 0)
+}
+
+// DoWriteRUStatistics write ru historical data into mysql.request_unit_by_group.
+func (r *RUStatsWriter) DoWriteRUStatistics(ctx context.Context) error {
+	// check if is already inserted
+	lastEndTime := GetLastExpectedTime(r.StartTime, r.Interval)
+	isInserted, err := r.islatestDataInserted(lastEndTime)
+	if err != nil {
+		return err
+	}
+	if isInserted {
+		logutil.BgLogger().Info("[ru_stats] ru data is already inserted, skip", zap.Stringer("end_time", lastEndTime))
+		return nil
+	}
+
+	lastStats, err := r.loadLatestRUStats()
+	if err != nil {
+		return err
+	}
+	needFetchData := true
+	if lastStats != nil && lastStats.Latest != nil {
+		needFetchData = lastStats.Latest.EndTime != lastEndTime
+	}
+
+	ruStats := lastStats
+	if needFetchData {
+		stats, err := r.fetchResourceGroupStats(ctx)
+		if err != nil {
+			return err
+		}
+		ruStats = &meta.RUStats{
+			Latest: &meta.DailyRUStats{
+				EndTime: lastEndTime,
+				Stats:   stats,
+			},
+		}
+		if lastStats != nil {
+			ruStats.Previous = lastStats.Latest
+		}
+		err = r.persistLatestRUStats(ruStats)
+		if err != nil {
+			return err
+		}
+	}
+	return r.insertRUStats(ruStats)
+}
+
+func (r *RUStatsWriter) fetchResourceGroupStats(ctx context.Context) ([]meta.GroupRUStats, error) {
+	groups, err := r.RMClient.ListResourceGroups(ctx, pd.WithRUStats)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	infos := r.InfoCache.GetLatest()
+	res := make([]meta.GroupRUStats, 0, len(groups))
+	for _, g := range groups {
+		groupInfo, exists := infos.ResourceGroupByName(model.NewCIStr(g.Name))
+		if !exists {
+			continue
+		}
+		res = append(res, meta.GroupRUStats{
+			ID:            groupInfo.ID,
+			Name:          groupInfo.Name.O,
+			RUConsumption: g.RUStats,
+		})
+	}
+	return res, nil
+}
+
+func (r *RUStatsWriter) loadLatestRUStats() (*meta.RUStats, error) {
+	snapshot := r.store.GetSnapshot(kv.MaxVersion)
+	metaStore := meta.NewSnapshotMeta(snapshot)
+	return metaStore.GetRUStats()
+}
+
+func (r *RUStatsWriter) persistLatestRUStats(stats *meta.RUStats) error {
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnOthers)
+	return kv.RunInNewTxn(ctx, r.store, true, func(ctx context.Context, txn kv.Transaction) error {
+		return meta.NewMeta(txn).SetRUStats(stats)
+	})
+}
+
+func (r *RUStatsWriter) islatestDataInserted(lastEndTime time.Time) (bool, error) {
+	end := lastEndTime.Format(time.DateTime)
+	start := lastEndTime.Add(-ruStatsInterval).Format(time.DateTime)
+	rows, sqlErr := execRestrictedSQL(r.sessPool, "SELECT 1 from mysql.request_unit_by_group where start_time = %? and end_time = %? limit 1", []interface{}{start, end})
```

Comment:
I have some questions about the logic of time judgment in the current implementation. If the writing delay, etc., is likely to occur a time difference greater than 1s, so this logic does not seem to work well

Reply:
Not really, you can see the inserted end_time is the `lastEndTime`, which is the expected end time of yesterday （for 24h duration）not matter what time this task is executed.